README.md
# time-logger

[![npm version](https://img.shields.io/npm/v/time-logger.svg?style=flat-square)](https://www.npmjs.com/package/time-logger)
[![travis build](https://img.shields.io/travis/gkjohnson/time-logger-js.svg?style=flat-square)](https://travis-ci.org/gkjohnson/time-logger-js)
[![lgtm code quality](https://img.shields.io/lgtm/grade/javascript/g/gkjohnson/time-logger-js.svg?style=flat-square&label=code-quality)](https://lgtm.com/projects/g/gkjohnson/time-logger-js/)

Utility library for tracking min / max / average evaluation time and tallying number of calls to code a code block in javascript. A more robust version of `console.time` and `console.timeEnd`.

## Use
```javascript
// in node
const TimeLogger = require('time-logger')

TimeLogger.begin('timing-label')
// ... code that needs to be profiled
TimeLogger.end('timing-label')
TimeLogger.dump('timing-label')
TimeLogger.clear('timing-label')
```

The above will print out the average, min, max, and call count of the code wrapped by `timing-label`:

```
timing-label         calls: 2        avg: 1532.5ms                  min: 1416.23ms                    max: 1649.1ms
```

Calling `dump()` and `clear()` with no arguments will print and clear _all_ accumulated timing metrics

## API
### getTime()

Returns the current time with highest precision method available (described below).

### begin(key)

Marks the beginning of the code block to track. A warning is printed and no changes are made if the provided key is already being tracked.

#### key : String

The identifier to use for the given code block.

### end(key)

Marks the end of the block of code to track timing and tallies for. A warning is printed if a corresponding `begin` call has not been made.

#### key : String

The identifier of the code block to end.

### clear(key = null)

Clears both pending and cached information for the provided key so timing can be started fresh.

#### key : String

The identifier to clear tracked information for. If null then all tracked and pending data is cleared.

### dump(key = null, clear = false)

Prints out timing information associated with provided key.

#### key : String

The identifier to print the information for. If null then all information and pending data is cleared.

#### clear : Boolean

Whether or not to call "clear" on the key after printing.

### getTimingData(key, createCopy = true) : Object

Returns the data associated with the provided key in the form of an object:

```js
{

    // the average time spent evaluating the code block
    avg: number,

    // the minimum and maximum amount of time
    min: number,
    max: number,

    // the number of times a code block was executed
    tally: number,

}
```

#### key : String

The identifier to get the associated data for.

#### createCopy : Boolean

Whether or not to create a _new_ object with the timing and tally data. If `false` then it is expected that this object is _not_ modified because it is used internally.

### getAllTimingData(createCopy = true) : Object

Returns all available timing data in object form.

#### createCopy

See `getTimingData`.

### getPending() : Array

Returns an array of all identifiers that have not had "end" called.

## Precision

The most precise timing measurements available are used depending on the platform. In browsers, `performance.now()` is used to get timing data, in Node `process.hrtime()` is used, and in Arangodb's Foxx `require('internal').time` is used.

When a precision timing function is unavailable, `Date.now()` is used, which gives coarse, millisecond precision.


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [0.3.1] - 2018-08-23
### Updated
- Documentation

## [0.3.0] - 2018-08-15
### Changed
- Remove `getTalliedMarks` function
- Rename `start` to `begin`
- Rename `getMarkData` to `getTimingData`
- Rename `getPendingMarks` to `getPending`

### Added
- Add `createCopy` argument to `getMarkData`
- Add `getAllTimingData` function

## [0.2.2] - 2018-08-15
### Fixed
- Function mistype in `getPendingMarks`


test/test.js
/* global
    describe it expect beforeEach
*/
const TimeLogger = require('../time-logger.js');

const originalConsole = { log: global.console.log, warn: global.console.warn };

describe('TimeLogger', () => {

    beforeEach(() => {

        TimeLogger.clear();
        global.console.log = originalConsole.log;
        global.console.warn = originalConsole.warn;

    });

    describe('begin / end', () => {

        it('should track time between calls', done => {

            TimeLogger.begin('label');

            setTimeout(() => {
                TimeLogger.end('label');

                const res = TimeLogger.getTimingData('label');
                expect(res.tally).toEqual(1);
                expect(Math.abs(1 - res.avg / 1000)).toBeLessThan(0.025);
                expect(res.avg).toEqual(res.max);
                expect(res.avg).toEqual(res.min);

                done();

            }, 1000);

        });

        it('should track multiple calls', () => {

            TimeLogger.begin('label');
            TimeLogger.end('label');

            TimeLogger.begin('label');
            TimeLogger.end('label');

            TimeLogger.begin('label');
            TimeLogger.end('label');

            const res = TimeLogger.getTimingData('label');
            expect(res.tally).toEqual(3);

        });

        it('should warn if end is called twice', () => {

            let called = 0;
            console.warn = msg => {

                called++;
                expect(/TimeLogger.end : 'end' called more than once/.test(msg)).toEqual(true);

            };

            TimeLogger.begin('label');
            TimeLogger.end('label');
            TimeLogger.end('label');

            expect(called).toEqual(1);
            expect(TimeLogger.getTimingData('label').tally).toEqual(1);

        });

        it('should warn if end is called with no begin', () => {

            let called = 0;
            console.warn = msg => {

                called++;
                expect(/TimeLogger.end : 'begin' not/.test(msg)).toEqual(true);

            };

            TimeLogger.end('label');

            expect(called).toEqual(1);

        });

        it('should warn if begom is called twice', () => {

            let called = 0;
            console.warn = msg => {

                called++;
                expect(/TimeLogger.begin : 'label' is already/.test(msg)).toEqual(true);

            };

            TimeLogger.begin('label');
            TimeLogger.begin('label');

            expect(called).toEqual(1);

        });

    });

    describe('dump', () => {

        it('should print data for the specified key', () => {

            let called = 0;
            global.console.log = (key, tally, avg, min, max) => {

                called++;
                expect(key).toBeTruthy();
                expect(tally).toBeTruthy();
                expect(avg).toBeTruthy();
                expect(min).toBeTruthy();
                expect(max).toBeTruthy();

                expect(key.trim()).toEqual('label');
                expect(/calls:[\s\d]+/.test(tally.trim())).toEqual(true);
                expect(/avg:[.\s\d]+ms/.test(avg.trim())).toEqual(true);
                expect(/min:[.\s\d]+ms/.test(min.trim())).toEqual(true);
                expect(/max:[.\s\d]+ms/.test(max.trim())).toEqual(true);

            };

            TimeLogger.begin('label');
            TimeLogger.end('label');
            TimeLogger.begin('label2');
            TimeLogger.end('label2');

            TimeLogger.dump('label');

            expect(called).toEqual(1);

        });

        it('should print all data if no key is passed', () => {

            let called = 0;
            global.console.log = () => called++;

            TimeLogger.begin('label');
            TimeLogger.end('label');
            TimeLogger.begin('label2');
            TimeLogger.end('label2');

            TimeLogger.dump();

            expect(called).toEqual(2);

        });

        it('should clear data if the appropriate parameter is passed', () => {

            TimeLogger.begin('label');
            TimeLogger.end('label');

            TimeLogger.dump('label', true);

            expect(Object.keys(TimeLogger.getAllTimingData()).length).toEqual(0);

        });

        it('should warn that an invalid key was passed', () => {

            let called = 0;
            global.console.warn = (msg, p1, p2, p3) => {

                called++;

                expect(/TimeLogger.dump : no timing details/.test(msg)).toEqual(true);
                expect(p1).toEqual(undefined);
                expect(p2).toEqual(undefined);
                expect(p3).toEqual(undefined);

            };

            TimeLogger.dump('none');

            expect(called).toEqual(1);

        });

    });

    describe('clear', () => {

        it('should clear only the logs of the label passed to it', () => {

            TimeLogger.begin('label');
            TimeLogger.end('label');

            TimeLogger.begin('label2');
            TimeLogger.end('label2');

            expect(TimeLogger.getTimingData('label')).toBeTruthy();
            expect(TimeLogger.getTimingData('label2')).toBeTruthy();

            TimeLogger.clear('label');

            expect(TimeLogger.getTimingData('label')).not.toBeTruthy();
            expect(TimeLogger.getTimingData('label2')).toBeTruthy();

        });

        it('should clear all logs if nothing is passed to it', () => {

            TimeLogger.begin('label');
            TimeLogger.end('label');

            TimeLogger.begin('label2');
            TimeLogger.end('label2');

            expect(TimeLogger.getTimingData('label')).toBeTruthy();
            expect(TimeLogger.getTimingData('label2')).toBeTruthy();

            TimeLogger.clear();

            expect(TimeLogger.getTimingData('label')).not.toBeTruthy();
            expect(TimeLogger.getTimingData('label2')).not.toBeTruthy();

        });

    });

    describe('getTimingData', () => {

        it('should have the expected keys', () => {

            TimeLogger.begin('label');
            TimeLogger.end('label');

            const res = TimeLogger.getTimingData('label');
            expect(res).toHaveProperty('avg');
            expect(res).toHaveProperty('min');
            expect(res).toHaveProperty('max');
            expect(res).toHaveProperty('tally');
            expect(Object.keys(res).length).toEqual(4);

        });

        it('should return a copy if return copy is truthy', () => {

            TimeLogger.begin('label');
            TimeLogger.end('label');

            const res1 = TimeLogger.getTimingData('label', true);
            const res2 = TimeLogger.getTimingData('label', true);

            expect(res1).not.toBe(res2);

        });

        it('should not return a copy if return copy is falsey', () => {

            TimeLogger.begin('label');
            TimeLogger.end('label');

            const res1 = TimeLogger.getTimingData('label', false);
            const res2 = TimeLogger.getTimingData('label', false);

            expect(res1).toBe(res2);

        });

        it('should return null if an invalid key is passed', () => {

            expect(TimeLogger.getTimingData('none')).toEqual(null);

        });

    });

    describe('getAllTimingData', () => {

        it('should return all stored timing data', () => {

            TimeLogger.begin('label');
            TimeLogger.end('label');
            TimeLogger.begin('label2');
            TimeLogger.end('label2');

            const res = TimeLogger.getAllTimingData();
            expect(Object.keys(res).length).toEqual(2);
            expect(res).toHaveProperty('label');
            expect(res.label).toHaveProperty('avg');
            expect(res.label).toHaveProperty('min');
            expect(res.label).toHaveProperty('max');
            expect(res.label).toHaveProperty('tally');

            expect(res).toHaveProperty('label2');

        });

        it('should return a copy if the parameter is truthy', () => {

            TimeLogger.begin('label');
            TimeLogger.end('label');

            const res1 = TimeLogger.getAllTimingData(true);
            const res2 = TimeLogger.getAllTimingData(true);

            expect(res1).not.toBe(res2);
            expect(res1.label).not.toBe(res2.label);

        });

        it('should not return a copy if the parameter is falsey', () => {

            TimeLogger.begin('label');
            TimeLogger.end('label');

            const res1 = TimeLogger.getAllTimingData(false);
            const res2 = TimeLogger.getAllTimingData(false);

            expect(res1).toBe(res2);
            expect(res1.label).toBe(res2.label);

        });

    });

    describe('getPending', () => {

        it('should contain labels that have not had `end` called', () => {

            TimeLogger.begin('label');
            TimeLogger.begin('label2');

            TimeLogger.begin('label3');
            TimeLogger.end('label3');
            TimeLogger.begin('label3');

            expect(TimeLogger.getPending().sort()).toEqual(['label', 'label2', 'label3']);

        });

        it('should not contain labels that have had `end` called', () => {

            TimeLogger.begin('label');
            TimeLogger.begin('label2');
            TimeLogger.end('label2');

            expect(TimeLogger.getPending()).toEqual(['label']);

        });

    });

});


time-logger.js
(exports => {

    /* Get Time Function */
    let getTime = null;

    // Browser
    if (getTime === null && typeof performance !== 'undefined' && performance.now) {

        getTime = () => performance.now();

        console.log('Using "performance.now" for timing');

    }

    // Node.js
    // https://stackoverflow.com/questions/23003252/window-performance-now-equivalent-in-nodejs
    if (getTime === null && typeof process !== 'undefined' && process.hrtime) {

        const startTime = process.hrtime();
        getTime = () => {

            const delta = process.hrtime(startTime);
            return delta[0] * 1e3 + delta[1] * 1e-6;

        };

        console.log('Using "process.hrtime" for timing');

    }

    // ArangoDB Foxx
    if (getTime === null && typeof require !== 'undefined') {

        try {

            // Using try catch in case the internal package
            // is made unavailable
            // internal.time() returns seconds since
            // epoch with microsecond timing
            const internal = require('internal');
            const getTimeNow = () => internal.time() * 1000; // ms
            const startTime = getTimeNow();
            getTime = () => getTimeNow() - startTime;

            console.log(`Using "require('internal').time" for timing`);

        } catch (e) {}
    }

    // Fallback to Date.now()
    if (getTime === null) {

        const startTime = Date.now();
        getTime = () => Date.now() - startTime;

        console.warn('Precise timing not available, falling back to millisecond precision with "Date.now()"');

    }

    /* Utilities */
    const pad = (str, width) => str.length < width ? pad(str + ' ', width) : str;

    /* Public API */
    // exposed mark fields for extra debug-ability
    exports._marks = {};
    exports._pendingMarks = {};

    // returns the time
    exports.getTime = () => getTime();

    // Begins a timing mark
    exports.begin = function(str) {

        if (str in this._pendingMarks) {

            console.warn(`TimeLogger.begin : '${ str }' is already being tracked. Call 'end' first.`);
            return;

        }

        this._pendingMarks[str] = this.getTime();

    };

    // Ends a timing mark
    exports.end = function(str) {

        if (!(str in this._pendingMarks)) {

            if (str in this._marks) console.warn(`TimeLogger.end : 'end' called more than once for 'begin' on '${ str }'.`);
            else console.warn(`TimeLogger.end : 'begin' not called for '${ str }'.`);
            return;

        }

        const delta = this.getTime() - this._pendingMarks[str];
        const details =
            this._marks[str] || {
                avg: 0,
                min: delta,
                max: delta,
                tally: 0,
            };

        details.tally++;
        details.avg += (delta - details.avg) / details.tally;
        details.min = Math.min(details.min, delta);
        details.max = Math.max(details.max, delta);
        this._marks[str] = details;

        delete this._pendingMarks[str];

        return delta;

    };

    // Clears out all accumulated timing
    exports.clear = function(str = null) {

        if (str != null) {

            if (str in this._marks) delete this._marks[str];
            if (str in this._pendingMarks) delete this._pendingMarks[str];

        } else {

            for (const key in this._marks) this.clear(key);
            for (const key in this._pendingMarks) this.clear(key);

        }

    };

    // Prints out all the logs for the given timing
    exports.dump = function(str = null, clear = false) {

        if (str != null) {

            if (str in this._marks) {

                const details = this._marks[str];

                console.log(
                    pad(str, 25),
                    pad(`calls: ${ details.tally }`, 10),
                    pad(`avg: ${ parseFloat(details.avg.toFixed(4)) }ms`, 20),
                    pad(`min: ${ parseFloat(details.min.toFixed(4)) }ms`, 20),
                    pad(`max: ${ parseFloat(details.max.toFixed(4)) }ms`, 20)
                );

                if (clear) this.clear(str);

            } else {

                console.warn(`TimeLogger.dump : no timing details to dump for '${ str }'`);

            }

        } else {

            for (const key in this._marks) this.dump(key, clear);

        }

    };

    // accessors for reading out timing data
    exports.getPending = function() {

        return Object.keys(this._pendingMarks);

    };

    exports.getAllTimingData = function(createCopy = true) {

        if (createCopy) {

            const res = {};
            for (var mark in this._marks) {

                res[mark] = this.getTimingData(mark, true);

            }

            return res;

        } else {

            return this._marks;

        }

    };

    exports.getTimingData = function(str, createCopy = true) {

        if (createCopy) {

            return str in this._marks ? Object.assign({}, this._marks[str]) : null;

        } else {

            return this._marks[str] || null;

        }

    };

})(typeof window !== 'undefined' ? window.TimeLogger = {} : module.exports = {});


perf-profiler.js
(exports => {
    /* Get Time Function */
    let getTime = null

    // Browser
    if (getTime === null && typeof performance !== 'undefined' && performance.now) {
        getTime = () => performance.now()

        console.log('Using "performance.now" for timing')
    }
    
    // Node.js
    // https://stackoverflow.com/questions/23003252/window-performance-now-equivalent-in-nodejs
    if (getTime === null && typeof process !== 'undefined' && process.hrtime) {
        const startTime = process.hrtime()
        getTime = () => {
            const delta = process.hrtime(startTime)
            return delta[0] * 1e3 + delta[1] * 1e-6
        }

        console.log('Using "process.hrtime" for timing')
    } 

    // ArangoDB Foxx
    if (getTime === null && typeof require !== 'undefined') {
        try {
            // Using try catch in case the internal package
            // is made unavailable
            // internal.time() returns seconds since
            // epoch with microsecond timing
            const internal = require('internal')
            const getTimeNow = () => internal.time * 1000 // ms
            const startTime = getTimeNow()
            getTime = () => getTimeNow() - startTime
            
            console.log(`Using "require('internal').time" for timing`)
        } catch (e) {}
    }

    // Fallback to Date.now()
    if (getTime === null) {
        const startTime = Date.now()
        getTime = () => Date.now() - startTime

        console.warn('Precise timing not available, falling back to millisecond precision with "Date.now()"')
    }

    /* Utilities */
    const pad = (str, width) => str.length < width ? pad(str + ' ', width) : str

    /* Public API */
    // exposed mark fields for extra debug-ability
    exports._marks = {}
    exports._pendingMarks = {}

    // returns the time
    exports.getTime = () => getTime()

    // Begins a timing mark
    exports.start = function(str) {
        this._pendingMarks[str] = this.getTime()
    }

    // Ends a timing mark
    exports.end = function(str) {
        if (!(str in this._pendingMarks)) {
            if (str in this._marks) console.warn(`'end' called more than once for 'start' on '${str}'`)
            else console.warn(`'start' not called for '${str}'`)
            return
        }

        const delta = this.getTime() - this._pendingMarks[str]
        const details = 
            this._marks[str] || {
                avg: 0,
                min: delta,
                max: delta,
                tally: 0
            }

        details.tally++
        details.avg += (delta - details.avg) / details.tally
        details.min = Math.min(details.min, delta)
        details.max = Math.max(details.max, delta)
        this._marks[str] = details

        delete this._pendingMarks[str]

        return delta
    }

    // Clears out all accumulated timing
    exports.clear = function(str = null) {
        if (str != null) {
            if (str in this._marks)           delete this._marks[str]
            if (str in this._pendingMarks)    delete this._pendingMarks[str]
        } else {
            for (let key in this._marks) this.clear(key)
            for (let key in this._pendingMarks) this.clear(key)
        }
    }

    // Prints out all the logs for the given timing
    exports.dump = function(str = null, clear = false) {
        if (str != null) {
            if (str in this._marks) {
                const details = this._marks[str]
                this.log(str, details)

                if (clear) this.clear(str)
            } else {
                console.warn(`no timing details to dump for '${str}'`)
            }
        } else {
            for (let key in this._marks) this.dump(key, clear)
        }
    }

    // Overridable Log function
    exports.log = (key, details) => {
        console.log(
            pad(key, 20),
            pad(`calls: ${details.tally}`, 15),
            pad(`avg: ${details.avg}ms`, 30),
            pad(`min: ${details.min}ms`, 30),
            pad(`max: ${details.max}ms`, 30))
    }

})(typeof window !== 'undefined' ? window.PerfProfiler = {} : module.exports = {})